tools/codemod/src/utils/TransformWidget.ts

```diff
@@ -0,0 +1,135 @@
+import { JSCodeshift, ASTPath, JSXElement, MemberExpression, ObjectExpression, ObjectProperty, Collection } from "jscodeshift";
+import { ObjectExpressionCollection, useObjectExpression } from "./ObjectExpression";
+import { ConfigExpression, getObjectProperty, jsxElementToUniFormat, uniFormatToObjectExpression } from "./TransformUtils";
+
+export function transformWidget(j: JSCodeshift, widget: ASTPath<JSXElement>): ConfigExpression { // TODO: registerMethod
+  const uni = jsxElementToUniFormat(j, widget);
+  const widgetObj = uniFormatToObjectExpression(j, uni);
+
+  // Add ---------------------------
+  // labelKey <- labelKey
+  // tooltipKey <- tooltipKey
+  // preferredPanelSize <- preferredPanelSize
+  // id <- id
+  // allowedPanels <- allowedPanelTargets
+  // badge <- badgeType
+  // canPopout <- canPopout
+  // canFloat <- isFloatingStateSupported / isFloatingStateWindowResizable, floatingContainerId, defaultFloatingPosition, defaultFloatingSize, hideWithUiWhenFloating
+  // defaultState <- defaultState
+  // content <- element
+  // icon <- iconSpec
+  // label <- label
+  // priority <- priority
+  // tooltip <- tooltip
+
+  // Ignore ------------------------
+  // onWidgetStateChanged
+  // saveTransientState
+  // restoreTransientState
+  // internalData
+  // applicationData
+  // control
+  // isFreeform
+  // isToolSettings
+  // isStatusBar
+  // fillZone
+  // syncEventIds
+  // stateFunc
+
+  replaceWidgetObjectExpression(j, j(widgetObj));
+  return { ...widgetObj, configName: uni.name };
+}
+
+export function replaceWidgetObjectExpression(j: JSCodeshift, widget: Collection<ObjectExpression>) {
+  useObjectExpression(j);
+  const objectToProperties = new Map<ASTPath<ObjectExpression>, ObjectProperty[]>();
+  (widget as ObjectExpressionCollection)
+    .removeProperty("key", (_path, property) => {
+      const idProp = getObjectProperty(j, _path.node, "id");
+      if (idProp === undefined && property.type === "ObjectProperty") {
+        const newIdProp = j.objectProperty(j.identifier("id"), property.value);
+        _path.node.properties.push(newIdProp);
+      }
+    })
+    .removeProperty("onWidgetStateChanged")
+    .removeProperty("saveTransientState")
+    .removeProperty("restoreTransientState")
+    .removeProperty("internalData")
+    .removeProperty("applicationData")
+    .removeProperty("control")
+    .removeProperty("isFreeform")
+    .removeProperty("isToolSettings")
+    .removeProperty("isStatusBar")
+    .removeProperty("fillZone")
+    .removeProperty("syncEventIds")
+    .removeProperty("stateFunc")
+    .renameProperty("iconSpec", "icon")
+    .renameProperty("isFloatingStateSupported", "canFloat")
+    .renameProperty("badgeType", "badge")
+    .renameProperty("element", "content")
+    .renameProperty("allowedPanelTargets", "allowedPanels", (_path, property) => {
```

Comment:
Very similar to `widget` transform, can we reuse this?

Reply:
The only reusable part here is the construction of `canFloat` property anything else would break the use of fluent api here and just bring more confusion as to what attributes `Widget` transformer handles

Reply:
All `Widget` properties are a subset of `WidgetConfig` properties, not only `canFloat`? Renames/removals are duplicated?